### PR TITLE
sweethome3d: fix impurity

### LINKED ISSUE #66471: sweethome3d impure? copied package don't work
**Describe the bug**
I copied a closure from one system (19.03) to a USB thumb drive and from that to another system (19.09), but the program don't work.

I think nix packages should be pure and work also on another system running a different NixOS channel. (Right?)

**To Reproduce**
1. have sweethome3d installed
2. plug in thumb drive
3. mount thumb drive (i use pantheons files app and open a terminal from there, then `pwd`)
4. create a folder on the thumb drive. mine is "nixpkgs"
5. copy to that folder: `nix copy --to file:///run/media/davidak/9AEE-1F0D/nixpkgs $(type -p sweethome3d)`
6. get the nix store path of the program:

```
[davidak@X230:~]$ type -p sweethome3d
/home/davidak/.nix-profile/bin/sweethome3d

[davidak@X230:~]$ ls -la /home/davidak/.nix-profile/bin/sweethome3d
lrwxrwxrwx 1 root root 87  1. Jan 1970  /home/davidak/.nix-profile/bin/sweethome3d -> /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d
```
7. plug the usb drive into another system and mount it, get the path
8. copy the nix store path to your system: `nix copy --from file:///run/media/davidak/9AEE-1F0D/nixpkgs /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
9.a you can install it for the current user if you like: `nix-env -i /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
9.b or just open a nix shell: `nix run /nix/store/hxp8riy9lzv5razfywxqgrqfj1db2vj7-sweethome3d-application-6.0/bin/sweethome3d`
10. now run the program:

```
$ sweethome3d 
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.eteks.sweethome3d.SweetHome3DBootstrap.main(Unknown Source)
Caused by: java.lang.UnsatisfiedLinkError: /tmp/extension4577508567935158164.so: libXxf86vm.so.1: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
	at java.lang.Runtime.load0(Runtime.java:809)
	at java.lang.System.load(System.java:1086)
	at com.jogamp.common.jvm.JNILibLoaderBase.loadLibraryInternal(JNILibLoaderBase.java:603)
	at com.jogamp.common.jvm.JNILibLoaderBase.access$000(JNILibLoaderBase.java:63)
	at com.jogamp.common.jvm.JNILibLoaderBase$DefaultAction.loadLibrary(JNILibLoaderBase.java:106)
	at com.jogamp.common.jvm.JNILibLoaderBase.loadLibrary(JNILibLoaderBase.java:487)
	at jogamp.nativewindow.NWJNILibLoader.access$000(NWJNILibLoader.java:39)
	at jogamp.nativewindow.NWJNILibLoader$1.run(NWJNILibLoader.java:49)
	at jogamp.nativewindow.NWJNILibLoader$1.run(NWJNILibLoader.java:41)
	at java.security.AccessController.doPrivileged(Native Method)
	at jogamp.nativewindow.NWJNILibLoader.loadNativeWindow(NWJNILibLoader.java:41)
	at jogamp.nativewindow.jawt.JAWTUtil.<clinit>(JAWTUtil.java:336)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.jogamp.nativewindow.NativeWindowFactory$3.run(NativeWindowFactory.java:344)
	at com.jogamp.nativewindow.NativeWindowFactory$3.run(NativeWindowFactory.java:340)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.jogamp.nativewindow.NativeWindowFactory.initSingleton(NativeWindowFactory.java:340)
	at com.jogamp.opengl.GLProfile.initProfilesForDefaultDevices(GLProfile.java:1721)
	at com.jogamp.opengl.GLProfile.access$000(GLProfile.java:80)
	at com.jogamp.opengl.GLProfile$1.run(GLProfile.java:230)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.jogamp.opengl.GLProfile.initSingleton(GLProfile.java:216)
	at com.jogamp.opengl.GLProfile.getProfileMap(GLProfile.java:2297)
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:1021)
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:1050)
	at com.jogamp.opengl.GLProfile.getMaxFixedFunc(GLProfile.java:803)
	at javax.media.j3d.JoglPipeline.initialize(JoglPipeline.java:131)
	at javax.media.j3d.Pipeline.createPipeline(Pipeline.java:92)
	at javax.media.j3d.MasterControl.loadLibraries(MasterControl.java:858)
	at javax.media.j3d.VirtualUniverse.<clinit>(VirtualUniverse.java:267)
	at javax.media.j3d.GraphicsConfigTemplate3D.isGraphicsConfigSupported(GraphicsConfigTemplate3D.java:348)
	at com.eteks.sweethome3d.j3d.Component3DManager.createGraphicsConfigurationTemplate3D(Unknown Source)
	at com.eteks.sweethome3d.j3d.Component3DManager.<init>(Unknown Source)
	at com.eteks.sweethome3d.j3d.Component3DManager.getInstance(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.addComponent3DRenderingErrorObserver(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.init(Unknown Source)
	at com.eteks.sweethome3d.SweetHome3D.main(Unknown Source)
	... 5 more
^C
```

**Expected behavior**
Program just work.

**Additional context**
I moved to a new flat with most of my computers and want to plan where to put my furniture. I installed sweethome3d on my notebook, but it is not powerfull enough. So i want to copy the program to my iMac. I have internet from my tethered mobile phone, so i can't just install it from the internet! sweethome3d is about 500 MB, that would be 1/4 of my data plan!!!

**Metadata**
X230:
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.63, NixOS, 19.03.173221.9ca57dc9171 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(davidak): `""`
 - channels(root): `"nixos-19.03.173221.9ca57dc9171, nixos-hardware, nixos-unstable-19.09pre186857.239fffc90d7"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

imac:
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.63, NixOS, 19.09.git.7db3602 (Loris)`
hangs after that

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: sweethome3d
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by tpwrules**
The sweethome3d program ships pre-compiled shared libraries for graphics that
need to be `patchelf`ed to work properly. Once the libraries are compressed into
the final .jar, Nix can no longer see the path references so they are not
declared a runtime dependency and the program won't run on a system that didn't
build it and doesn't happen to have the right paths already present on it by
chance.

This patch switches to autoPatchelfHook to ensure the libraries get patched
correctly, then scans the libraries to make a plain-text file with the same
paths so that Nix can see the references and account for them. I suspect this
problem might be lurking in other Java programs in nixpkgs.

Fixes #66471.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mindavi** [comment]
See https://github.com/NixOS/nixpkgs/pull/171027 for prior art. The way it's done here might be better since it's not based on a static list of dependencies someone some day compiled. What do you think @trofi?

File: pkgs/applications/misc/sweethome3d/default.nix [hunk 1]
```diff
@@ -80,6 +84,10 @@ let
         --prefix XDG_DATA_DIRS : "$XDG_ICON_DIRS:${gtk3.out}/share:${gsettings-desktop-schemas}/share:$out/share:$GSETTINGS_SCHEMAS_PATH" \
         --add-flags "-Dsun.java2d.opengl=true -jar $out/share/java/${module}-${version}.jar -cp $out/share/java/Furniture.jar:$out/share/java/Textures.jar:$out/share/java/Help.jar -d${toString stdenv.hostPlatform.parsed.cpu.bits}"
 
+
+      # remember the store paths found inside the .jar libraries
+      cp .jar-paths $out
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/sweethome3d/default.nix:89]: Can we find a better place for the file? Otherwise it ends up in ``/``.

Comment (tpwrules) [reply pkgs/applications/misc/sweethome3d/default.nix:89]: Would you be good with `/nix-support/depends`? I see it on #46141 and Bazel already uses it for this same purpose: https://github.com/NixOS/nixpkgs/blob/af56903adb67202dbca2a8b31d3b162b0ee71cd6/pkgs/development/tools/build-managers/bazel/bazel_0_26/default.nix#L379-L386

Comment (SuperSandro2000) [reply pkgs/applications/misc/sweethome3d/default.nix:89]: That sounds better to me if it works.

**Comment by tpwrules** [comment]
I checked and it looks very unlikely that `autoPatchelf` can handle PE or Mach-O files. But being mindful of other types of files store paths can be hiding in is wise.

I actually ran into this problem on something I can't share and there I had to unzip all the .jar files, `autoPatchelf` them, then re-zip them. So that might even be necessary in some cases.

**Review by trofi — APPROVED** [review]
Extracting store references automatically is nice.

Yeah, sounds like a new `fixupHook` would be useful for java packages (does not have to be a part of this PR). `find/grep` would probably turn into something  more robust (like considering `config.nix.storeDir`, or running external tool similar to `nix`'s `scanForReferences()`)

**Pull request merged**